Judge tool: Search Trace Regex

dbczumar opened the pull request:
### Related Issues/PRs

<!-- Uncomment 'Resolve' if this PR can close the linked items. -->
<!-- Resolve --> #xxx

### What changes are proposed in this pull request?

Judge tool: Search Trace Regex

### How is this PR tested?

- [ ] Existing unit/integration tests
- [X] New unit/integration tests
- [ ] Manual tests

<!-- Attach code, screenshot, video used for manual testing here. -->

### Does this PR require documentation update?

- [X] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Release Notes

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

<!-- Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change. -->

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [X] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/prompts`: MLflow prompt engineering features, prompt templates, and prompt management
- [ ] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [X] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

<!-- patch -->

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [X] No (this PR will be included in the next minor release)

dbczumar committed:
Add SearchTraceRegexTool for regex-based trace span searching

This commit adds a new judge tool that enables regex-based searching across
all spans in a trace. The tool searches through span inputs, outputs,
attributes, and names using case-insensitive regex patterns.

Key features:
- Case-insensitive regex search across all span fields
- Returns matches with surrounding context (100 chars before/after)
- Configurable match limits (default: 50)
- Comprehensive error handling for invalid regex patterns
- Integration with existing judge tools registry

Files added:
- mlflow/genai/judges/tools/search_trace_regex.py: Main tool implementation
- tests/genai/judges/test_search_trace_regex_tool.py: Function-based test suite (17 tests)

Files modified:
- mlflow/genai/judges/tools/constants.py: Added SEARCH_TRACE_REGEX constant
- mlflow/genai/judges/tools/__init__.py: Added exports for new tool
- mlflow/genai/judges/tools/registry.py: Registered new tool globally

All tests pass and code follows MLflow coding standards.

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Drastically simplify SearchTraceRegexTool implementation

Replace complex field-by-field search logic with simple JSON string search:
- Use trace.to_json() to get complete JSON representation
- Run regex directly on the JSON string
- Remove unnecessary span ID extraction complexity
- Use simple "trace" identifier for span_id in results

This reduces the search logic from 70+ lines to ~20 lines while maintaining
the same functionality. The approach is much more maintainable and less
error-prone.

All tests pass with the simplified implementation.

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Extract match creation logic into helper function

Factor out RegexMatch creation logic into _create_regex_match() helper:
- Reduces code duplication and improves readability
- Makes the main search loop cleaner and more focused
- Helper function handles context extraction and ellipses logic
- Maintains same functionality with better code organization

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Remove unnecessary error handling for empty traces

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Move _create_regex_match to be a class method below invoke

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Remove useless inline comments

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Update docstrings to accurately reflect tool searches entire trace, not just spans

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Replace 'all trace data' with 'all trace fields'

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Remove useless trace.data None check and clean up needless newlines

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Add missing type hint for match parameter

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Shorten test suite while maintaining coverage (391→229 lines)

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Group related test cases together (max matches tests now adjacent)

Signed-off-by: dbczumar <corey.zumar@databricks.com>

BenWilson2 commented on mlflow/genai/judges/tools/search_trace_regex.py:
```diff
@@ -0,0 +1,144 @@
+        """Create a RegexMatch with surrounding context from a regex match object."""
+        matched_text = match.group()
+        start, end = match.span()
+        context_start = max(0, start - 100)
```
Do you think that we should make this configurable or has your testing confirmed 100 characters to be an effective value?

BenWilson2 reviewed (approved):
tiny nit. Otherwise looks great!

dbczumar commented on tests/genai/judges/test_search_trace_regex_tool.py:
```diff
@@ -0,0 +1,246 @@
+)
+
+
+def create_test_trace():
```
Definitely, fixed!

dbczumar committed:
Make regex search context configurable and convert test to fixture

Address @BenWilson2's feedback:
- Add surrounding_content_length parameter to make context start/end offset configurable (default: 100)
- Convert create_test_trace() to pytest fixture for efficient test utilization
- Remove unnecessary variable assignments in tests

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Merge branch 'judges' into span_regex_search

Signed-off-by: Corey Zumar <39497902+dbczumar@users.noreply.github.com>
